azurelinuxagent/ga/telemetry_service.py

```diff
@@ -0,0 +1,146 @@
+# Microsoft Azure Linux Agent
+#
+# Copyright 2020 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+import datetime
+import threading
+import time
+import traceback
+
+from azurelinuxagent.common import logger
+from azurelinuxagent.common.event import add_event, WALAEventOperation
+from azurelinuxagent.common.exception import ServiceStoppedError
+from azurelinuxagent.common.future import ustr, Queue, Full
+from azurelinuxagent.common.interfaces import ThreadHandlerInterface
+
+
+def get_send_telemetry_events_handler(protocol_util):
+    return SendTelemetryEventsHandler(protocol_util)
+
+
+class SendTelemetryEventsHandler(ThreadHandlerInterface):
+    """
+    This Handler takes care of sending all telemetry out of the agent to Wireserver. It sends out data as soon as
+    there's any data available in the queue to send.
+    """
+
+    _THREAD_NAME = "SendTelemetryEventsHandler"
+    _MAX_TIMEOUT = datetime.timedelta(seconds=5).seconds
+    _MIN_EVENTS_TO_BATCH = 30
+    _MIN_BATCH_WAIT_TIME = datetime.timedelta(seconds=5)
+
+
+    def __init__(self, protocol_util):
+        self._protocol = protocol_util.get_protocol()
+        self.should_run = True
+        self._thread = None
+        self._should_process_events = threading.Event()
+        self._queue = Queue()
```

Comment:
let's continue our offline discussion about using an unbounded queue (no size limit)

Reply:
Synced offline, decided to leave it unbounded for now. Captured a Todo in our code for this gap.